File: 208.implement-trie-prefix-tree.go
/*
 * @lc app=leetcode id=208 lang=golang
 *
 * [208] Implement Trie (Prefix Tree)
 */

// @lc code=start
// S: O(n * 26), n := maximum word len
type Trie struct {
	children [26]*Trie
	isWord bool
}


func Constructor() Trie {
	return Trie{}
}


// T: O(len(word))
func (this *Trie) Insert(word string)  {
	node := this
	for _, c := range word {
		ci := c - 'a'
		if node.children[ci] == nil {
			node.children[ci] = &Trie{}
		}
		node = node.children[ci]
	}
	node.isWord = true
}


// T: O(len(word))
func (this *Trie) Search(word string) bool {
    node := this
	for _, c := range word {
		ci := c - 'a'
		if node.children[ci] == nil {
			return false
		}
		node = node.children[ci]
	}
	return node.isWord
}


// T: O(len(prefix))
func (this *Trie) StartsWith(prefix string) bool {
    node := this
	for _, c := range prefix {
		ci := c - 'a'
		if node.children[ci] == nil {
			return false
		}
		node = node.children[ci]
	}
	return true
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
// @lc code=end

File: 208.implement-trie-prefix-tree.cpp
/*
 * @lc app=leetcode id=208 lang=cpp
 *
 * [208] Implement Trie (Prefix Tree)
 */
#include <string>

using std::string;

// @lc code=start
struct Node {
    Node* next[26];
    bool isEnd;

    Node() : isEnd(false) {
        for (int i = 0; i < 26; i++) {
            next[i] = nullptr;
        }
    }
};

// T: O(l), l := len(word)
// S: O(l * alpha_size)
class Trie {
    Node *root;

public:
    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        Node *tmp = root;
        for (const auto &c : word) {
            int cidx = c - 'a';
            if (!tmp->next[cidx]) {
                tmp->next[cidx] = new Node();
            }
            tmp = tmp->next[cidx];
        }
        tmp->isEnd = true;
    }
    
    bool search(string word) {
        Node *tmp = root;
        for (const auto& c : word) {
            int cidx = c - 'a';
            if (!tmp->next[cidx]) {
                return false;
            }
            tmp = tmp->next[cidx];
        }
        return tmp->isEnd;
    }
    
    bool startsWith(string prefix) {
        Node *tmp = root;
        for (const auto& c : prefix) {
            int cidx = c - 'a';
            if (!tmp->next[cidx]) {
                return false;
            }
            tmp = tmp->next[cidx];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
// @lc code=end

